[InstCombine] Add combines/simplifications for `llvm.ptrmask`

goldsteinn opened the pull request:
- [InstSimplify] Add tests for simplify `llvm.ptrmask`; NFC
- [InstSimplify] Add some basic simplifications for `llvm.ptrmask`
- [InstCombine] Add tests for combining `llvm.ptrmask`; NFC
- [InstCombine] Deduce `align` and `nonnull` return attributes for `llvm.ptrmask`
- [InstCombine] Implement `SimplifyDemandedBits` for `llvm.ptrmask`
- [InstCombine] Preserve return attributes when merging `llvm.ptrmask`
- [InstCombine] Merge consecutive `llvm.ptrmask` with different mask types if a mask is constant.
- [InstCombine] Fold `(ptrtoint (ptrmask p0, m0))` -> `(and (ptrtoint p0), m0)`

goldsteinn committed:
[InstSimplify] Add tests for simplify `llvm.ptrmask`; NFC

Differential Revision: https://reviews.llvm.org/D156632

goldsteinn committed:
[InstSimplify] Add some basic simplifications for `llvm.ptrmask`

Mostly the same as `and`. We also have a check for a useless
`llvm.ptrmask` if the ptr is already known aligned.

Differential Revision: https://reviews.llvm.org/D156633

goldsteinn committed:
[InstCombine] Add tests for combining `llvm.ptrmask`; NFC

Differential Revision: https://reviews.llvm.org/D156634

goldsteinn committed:
[InstCombine] Preserve return attributes when merging `llvm.ptrmask`

If we have assosiated attributes i.e `([ret_attrs] (ptrmask (ptrmask
p0, m0), m1))` we should preserve `[ret_attrs]` when combining the two
`llvm.ptrmask`s.

Differential Revision: https://reviews.llvm.org/D156638

goldsteinn committed:
[InstCombine] Fold `(ptrtoint (ptrmask p0, m0))` -> `(and (ptrtoint p0), m0)`

`and` is generally more supported so if we have a `ptrmask` anyways
might as well use `and`.

Differential Revision: https://reviews.llvm.org/D156640

goldsteinn commented:
@nikic, think fixed all issues with `null`. Only case maybe not okay is for `llvm.ptrmask(undef, X)` -> `null`.

fhahn commented on llvm/test/Transforms/InstCombine/align-addr.ll:
```diff
@@ -183,8 +183,7 @@ define <16 x i8> @ptrmask_align8_ptr_align1(ptr align 1 %ptr) {
 ; TODO: Should be able to drop the ptrmask
```
remove TODO now?

fhahn commented on llvm/test/Transforms/InstCombine/align-addr.ll:
```diff
@@ -196,8 +195,7 @@ define <16 x i8> @ptrmask_align8_ptr_align8(ptr align 8 %ptr) {
 ; TODO: Should be able to drop the ptrmask
```
remove TODO now

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1964,18 +1964,78 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
+          NeedsNew = true;
+        }
+      }
+      if (InnerMask->getType() == Op1->getType()) {
```
`OkayToMerge` is unused?

nikic commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6397,6 +6397,48 @@ static Value *simplifyBinaryIntrinsic(Function *F, Value *Op0, Value *Op1,
+    // useless should be updated to use generic constants.
+    assert(!Op0->getType()->isVectorTy() && !Op1->getType()->isVectorTy() &&
+           "These simplifications where written at a time when ptrmask did not "
+           "support vector types and may not work for vectors");
```
As far as I can see all the code here is already compatible with vectors, so not sure this assert makes sense (and I don't understand what your TODO is referring to either).

goldsteinn replied:
I guess my feeling is since its untested, we shouldn't assume it works. Seems lower cost to just delete these lines when/if adding vec support than to potentially leave buggy codes.

nikic replied:
Let's make it testable then: https://github.com/llvm/llvm-project/pull/67434

nikic commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6397,6 +6397,48 @@ static Value *simplifyBinaryIntrinsic(Function *F, Value *Op0, Value *Op1,
+
+    // NOTE: We can't apply these simplifications based on the value of Op1
+    // because we need to preserve provenance.
+    if (isa<PoisonValue>(Op0))
```
You can also return poison if op1 is poison.

nikic commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6397,6 +6397,48 @@ static Value *simplifyBinaryIntrinsic(Function *F, Value *Op0, Value *Op1,
+      if (match(Op1, m_PtrToInt(m_Specific(Op0))))
+        return Op0;
+
+      // TODO: We may have attributes assosiated with the return value of the
```
associated

Though really, I think you can just drop this TODO. Trying to preserve ptrmask attributes is almost certainly not worthwhile.

goldsteinn replied:
How about "TODO(unimportant)" if someone is searching for why attributes are being lost, may help them debug.

goldsteinn replied:
I changed it to a note.

nikic commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6397,6 +6397,48 @@ static Value *simplifyBinaryIntrinsic(Function *F, Value *Op0, Value *Op1,
+      // llvm.ptrmask intrinsic that will be lost when we just return the
+      // operand. We should try to preserve them.
+      if (match(Op1, m_AllOnes()))
+        return Op0;
```
You can also handle undef op1 like -1 here.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1985,6 +1985,28 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
+      Changed = true;
+    }
+
+    // Known bits will capture if we had alignment information assosiated with
```
associated

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1985,6 +1985,28 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
+    // Known bits will capture if we had alignment information assosiated with
+    // the pointer argument.
+    if (Known.countMinTrailingZeros() > Log2(CI.getRetAlign().valueOrOne())) {
+      if (CI.hasRetAttr(Attribute::Alignment))
```
No need to call hasRetAttr

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1985,6 +1985,28 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
+        CI.removeRetAttr(Attribute::Alignment);
+      CI.addRetAttr(
+          Attribute::get(CI.getContext(), Attribute::Alignment,
+                         uint64_t(1) << Known.countMinTrailingZeros()));
```
Attribute::getWithAlignment(). Also you need to protect against MaximumAlignment here.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:
```diff
@@ -898,6 +910,53 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
+        assert(!LHSKnown.hasConflict() && "Bits known to be one AND zero?");
+
+        Known = LHSKnown & RHSKnown;
+        KnownBitsComputed = DemandedMask.isAllOnes();
```
This should always be true.

goldsteinn replied:
At the moment sure, but could imagine otherwise?

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:
```diff
@@ -898,6 +910,53 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
+        // the backend.
+        if (DemandedMask.isSubsetOf(Known.Zero))
+          return replaceOperand(
+              *I, 1, Constant::getNullValue(I->getOperand(1)->getType()));
```
It looks like this is untested (or the test updates are in the wrong commit).

goldsteinn replied:
Added proper tests.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1984,10 +1984,30 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
+          // Need to create a new one here, as the intrinsic id needs to change.
+          NeedsNew = true;
+        }
+      }
```
This is the wrong approach. You should be canonicalizing the second operand to match the pointer size instead and then let the usual folds handle the rest.

goldsteinn replied:
Done (only canonicalize constants, could do all masks though).

arichardson commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6397,6 +6397,48 @@ static Value *simplifyBinaryIntrinsic(Function *F, Value *Op0, Value *Op1,
+      // TODO: We may have attributes associated with the return value of the
+      // llvm.ptrmask intrinsic that will be lost when we just return the
+      // operand. We should try to preserve them.
+      if (match(Op1, m_AllOnes()) || Q.isUndefValue(Op1))
```
Is the argument to ptrmask defined as being sign-extended? If not this needs a bitwidth check.

nikic replied:
There is a bitwidth check above (you even commented on it ^^)

arichardson replied:
Sorry about the noise, my comment above made me miss the `{` so I assumed this was not nested.

nikic commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6397,6 +6397,48 @@ static Value *simplifyBinaryIntrinsic(Function *F, Value *Op0, Value *Op1,
+      return Constant::getNullValue(Op0->getType());
+
+    if (Op1->getType()->getScalarSizeInBits() ==
+        Q.DL.getPointerTypeSizeInBits(Op0->getType())) {
```
This is just following the currently LangRef definition (https://llvm.org/docs/LangRef.html#llvm-ptrmask-intrinsic), which refers to the pointer size, not the pointer index size. If you'd like to change the spec to use the pointer index size, then a LangRef patch would be the starting point.

arichardson replied:
Looking at the current langref:
> The result of ptrmask(ptr, mask) is equivalent to getelementptr ptr, (ptrtoint(ptr) & mask) - ptrtoint(ptr).

Since it's defined in terms of GEP it indicates to me that the masking cannot affect more than the index bits. GEP semantics:
>  If the offsets have a different width from the pointer’s index type, they are sign-extended or truncated to the width of the pointer’s index type.

goldsteinn replied:
@arichardson guess my feeling is for this series (just adding basic support) would prefer to keep checks on the more conservative side. Think its easier to expand than try and do it all at once.

arichardson replied:
Sure, I think keeping it minimal makes sense. I was just wondering if it's correct to use DL.getPointerIndexSizeInBits() here instead? This should be the same behaviour for almost all targets but would make this fold work for the downstream CHERI targets as well once merged.

nikic replied:
Drop this check now? For the following transforms:

 * The ptrtoint one is safe, but you won't actually be able to trigger it if index type is smaller.
 * The allones/undef ones are safe and can be tested.
 * The knownbits one is safe with a truncate of the known bits. Can be tested.

goldsteinn commented:
Okay, rebased ontop of your vec support patches + added vec support/tests.

goldsteinn commented:
ping.

arichardson commented:
Since the langref definition is based on GEP, can we use index width instead of pointer width? If you disagree with this, it would still be good if you could add a test where the datalayout has index width < pointer width to show the current semantics.

goldsteinn commented:
> Since the langref definition is based on GEP, can we use index width instead of pointer width? If you disagree with this, it would still be good if you could add a test where the datalayout has index width < pointer width to show the current semantics.

Any target inmind? But I think ptrwidth makes more sense given that the result is a pointer. It would be awkward to have resize things like knownbits / etc... and would somewhat make ptrmask useless as basically any mask would result in a low-bit only pointer.

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6397,6 +6397,41 @@ static Value *simplifyBinaryIntrinsic(Function *F, Value *Op0, Value *Op1,
+      Constant *C;
+      if (match(Op1, m_ImmConstant(C))) {
+        KnownBits PtrKnown =
+            computeKnownBits(Op0, Q.DL, /*Depth*/ 0, Q.AC, Q.CxtI, Q.DT);
```
Done

goldsteinn commented:
ping.

nikic commented:
We should update LangRef and require that the integer arg has the size of the pointer index type (with current GEP semantics implying a 1-extend to pointer type width), implement a Verifier check for that, and add an assert in SDAG lowering that the index type size is the pointer size -- I don't believe we currently have any in-tree SDAG support for pointers with narrow index types, so this is just a breadcrumb for out-of-tree targets to implement a 1-extend at that point.

I was planning to do this, but never found the time.

goldsteinn commented:
> We should update LangRef and require that the integer arg has the size of the pointer index type 
As in disallow say `ptrmask.p0.i32` on typical systems with 64-bit pointer/index?

> (with current GEP semantics implying a 1-extend to pointer type width), implement a Verifier check for that, and add an assert in SDAG lowering that the index type size is the pointer size -- I don't believe we currently have any in-tree SDAG support for pointers with narrow index types, so this is just a breadcrumb for out-of-tree targets to implement a 1-extend at that point.
> 
> I was planning to do this, but never found the time.

goldsteinn commented:
> > > We should update LangRef and require that the integer arg has the size of the pointer index type
> > 
> > 
> > As in disallow say `ptrmask.p0.i32` on typical systems with 64-bit pointer index? (So a lot of the current usages). Or something else?
> 
> Yes, exactly. I've started working on a patch at #69343, but it's not quite done yet.

Okay, Ill wait on LangRef update to get this in.
Do you want me to take up the langref patch or do you want to finish it up?
I'm happy to (would like to a bit in fact), although its in a few areas I'm not really familiar with (Verifier)
so will probably be slower going that what you can get done.

dtcxzyw commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6411,6 +6411,41 @@ static Value *simplifyBinaryIntrinsic(Function *F, Value *Op0, Value *Op1,
+
+      Constant *C;
+      if (match(Op1, m_ImmConstant(C))) {
+        KnownBits PtrKnown =
+            computeKnownBits(Op0, Q.DL, /*Depth=*/0, Q.AC, Q.CxtI, Q.DT);
```
```suggestion
        KnownBits PtrKnown = computeKnownBits(Op0, /*Depth=*/0, Q);
```
See also new ValueTracking helper functions introduced by be57381a4a08b0b6a89d5b5fdec0880b202e99f4.

goldsteinn replied:
Done

goldsteinn commented:
Rebased ontop of nikic's patch

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1973,6 +1978,28 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
+    // Known bits will capture if we had alignment information associated with
+    // the pointer argument.
+    if (NewAlignmentLog > Log2(CI.getRetAlign().valueOrOne())) {
+      CI.removeRetAttr(Attribute::Alignment);
```
No need to explicitly remove the old attribute.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:
```diff
@@ -143,7 +155,7 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
 
   KnownBits LHSKnown(BitWidth), RHSKnown(BitWidth);
-
+  bool MaySimplifyAsConstant = true;
```
Possibly this flag can be replaced by a isPointerTy check below? This seems like it should also hold for other instructions working on pointers, if we were to handle them.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:
```diff
@@ -984,7 +1041,7 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
   // If the client is only demanding bits that we know, return the known
   // constant.
-  if (DemandedMask.isSubsetOf(Known.Zero|Known.One))
+  if (MaySimplifyAsConstant && DemandedMask.isSubsetOf(Known.Zero | Known.One))
```
FWIW, it would be legal to return a inttoptr constant for non-zero values. But not sure if it would be a good idea, so staying conservative is reasonable.

goldsteinn replied:
Ill add that as a todo. Although maybe indicative we want to the flag to be based on the exact intrin as opposed to `isPointer` as different intrins may have different preference. For `@llvm.ptrmask` for example. we can easily represent the most common constant (`null`) by just zeroing the mask so there isn't much benefit to constant return. But for other intrins might be different.

goldsteinn replied:
Leaving as TODO + made just a pointer check.

arichardson commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6411,6 +6411,40 @@ static Value *simplifyBinaryIntrinsic(Function *F, Value *Op0, Value *Op1,
+    if (Q.isUndefValue(Op0) || match(Op0, m_Zero()))
+      return Constant::getNullValue(Op0->getType());
+
+    if (Op1->getType()->getScalarSizeInBits() ==
```
I believe this branch can be replaced with assert(Op1->getType()->getScalarSizeInBits() == Q.DL.getPointerIndexSizeInBits(Op0->getType()) since the verified now ensures the argument matches the index size since #69343

goldsteinn replied:
I think it needs to be pointer size (as opposed to index size) as we are checking for useless masks. I.e if pointer size > index size all 1s mask isn't redundant.

arichardson replied:
It is redundant since it does not affect more than the index bits (see clarification in the langref added in https://github.com/llvm/llvm-project/pull/69343)

goldsteinn committed:
[InstCombine] Implement `SimplifyDemandedBits` for `llvm.ptrmask`

Logic basically copies 'and' but we can't return a constant if the
result == rhs (mask) so that case is skipped.

goldsteinn committed:
[InstCombine] Deduce `align` and `nonnull` return attributes for `llvm.ptrmask`

We can deduce the former based on the mask / incoming pointer
alignment.  We can set the latter based if know the result in non-zero
(this is essentially just caching our analysis result).

Differential Revision: https://reviews.llvm.org/D156636

arichardson reviewed (commented):
All my comments have been resolved, thank you very much!

nikic commented:
The InstCombine/ptrmask.ll test is failing in CI.

goldsteinn commented:
> The InstCombine/ptrmask.ll test is failing in CI.

Bah, forgot to rebase the instcombine ptrmask tests ontop of simplication fix.

nikic commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -6411,6 +6411,44 @@ static Value *simplifyBinaryIntrinsic(Function *F, Value *Op0, Value *Op1,
+      KnownBits PtrKnown = computeKnownBits(Op0, /*Depth=*/0, Q);
+      // See if we only masking off bits we know are already zero due to
+      // alignment.
+      APInt IrrelivantPtrBits =
```
Irrelevant

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:
```diff
@@ -978,8 +1031,10 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
-  // constant.
-  if (DemandedMask.isSubsetOf(Known.Zero|Known.One))
+  // constant. We can't directly simplify pointers as a constant because of
+  // pointer provanance.
```
provenance

nikic reviewed (approved):
LGTM!
